Make Blocks depend on BOPS

tenderlove opened the pull request:
When a BOP is redefined, the BOP redefinition callback will invalidate
any blocks that depend on BOPS.  This allows us to eliminate runtime
checks for BOP redefinition.

tenderlove committed:
Make Blocks depend on BOPS

When a BOP is redefined, the BOP redefinition callback will invalidate
any blocks that depend on BOPS.  This allows us to eliminate runtime
checks for BOP redefinition.

maximecb commented:
We're segfaulting somehow, but not on all platforms?

maximecb commented:
@tenderlove When you get the chance, can you investigate the segfaults on this? :O

If it's helpful, we can pair Friday afternoon.

XrXr committed:
Check for null

Sometimes you bop before you init, you know?

maximecb commented on yjit_iface.c:
```diff
@@ -472,7 +472,9 @@ block_invalidation_iterator(st_data_t key, st_data_t value, st_data_t data) {
 rb_yjit_bop_redefined(VALUE klass, const rb_method_entry_t *me, enum ruby_basic_operators bop)
 {
-    st_foreach(blocks_assuming_bops, block_invalidation_iterator, 0);
+    if (blocks_assuming_bops) {
```
You are the real MVP

Commit:
Remove trailing whitespaces

maximecb commented:
Hmmm we're failing some tests but only on some platforms. This is fishy 🐟

XrXr committed:
Include ruby/ruby.h before vm_core.h

This is how vm.c does it, and if we don't follow it in yjit
compilation units, rb_vm_t would have a conflicting size.

XrXr committed:
Merge branch 'yjit' into bop-dependencies

XrXr committed:
Tweak include order in yjit_core.c

This is playing it safe. Ruby headers rely on some macros generated by
autoconf that affect what system headers behave. We should take care to
bring in those configs before including systme headers to stay
consistent with the rest of Ruby.
